### ISSUE TITLE
propEq type definition does not guarantees type safety

### ISSUE THREAD

**Issue by peterreisz** [selfrefactor/rambda#476]
The current definition accepts any as the last parameter

```ts
export function propEq<T>(propToFind: string | number, valueToMatch: T, obj: any): boolean;
export function propEq<T>(propToFind: string | number, valueToMatch: T): (obj: any) => boolean;
export function propEq(propToFind: string | number): {
  <T>(valueToMatch: T, obj: any): boolean;
  <T>(valueToMatch: T): (obj: any) => boolean;
};
```

The type safe solution would be something like this:

```ts
export function propEq<T, K extends keyof T>(propToFind: K, valueToMatch: T[K], obj: T): boolean;
export function propEq<T, K extends keyof T>(propToFind: K, valueToMatch: T[K]): (obj: T) => boolean;
export function propEq<T, K extends keyof T>(propToFind: K): {
   (valueToMatch: T[K], obj: T): boolean;
   (valueToMatch: T[K]): (obj: T) => boolean;
};
```

**Comment by selfrefactor** [comment]
Thank you  @peterreisz for pointing this out.

I copied the definitions from `@types/ramda` and I created a `dtslint` test to confirm the change. You can see it in the referenced commit.

**Comment by selfrefactor** [comment]
Released with `5.6.1`

**Comment by peterreisz** [comment]
@selfrefactor there is a problem with the `Record` solution when using optional properties.
When the type has optional property you cannot write code which compiles.
The built-in Record type constructs a new type, but in that type the prop will be required, which will not match with the given type.

```ts
interface MyType {
    optional?: string | number;
}

function propEqWithRecord<K extends string | number, V>
  (propToFind: K, valueToMatch: V, obj: Record<K, V>): boolean {
    return false;
}

function propEqWithoutRecord<K extends keyof T, T>
  (propToFind: K, valueToMatch: T[K], obj: T): boolean {
    return false;
}

const myObject: MyType = {};
const valueToFind = '1111';
const optionalValueToFind: string | number | undefined = '1111';

// These will not compile
propEqWithRecord('optional', valueToFind, myObject);
propEqWithRecord('optional', optionalValueToFind, myObject);

// These will compile
propEqWithoutRecord('optional', valueToFind, myObject);
propEqWithoutRecord('optional', optionalValueToFind, myObject);
```

**Comment by selfrefactor** [comment]
@peterreisz I will run `dtslint` test and most probably I will apply your suggestion.

**Comment by selfrefactor** [comment]
Released with `5.6.2`
